internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -851,7 +851,32 @@ func (r *BMACReconciler) SetupWithManager(mgr ctrl.Manager) error {
 			}}}
 		}
 
-		bmh, err := r.findBMH(ctx, agent)
+		bmh, err := r.findBMHByAgent(ctx, agent)
+		if bmh == nil || err != nil {
+			return []reconcile.Request{}
+		}
+
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Namespace: bmh.Namespace,
+			Name:      bmh.Name,
+		}}}
+	}
+
+	mapInfraEnvToBMH := func(a client.Object) []reconcile.Request {
+		ctx := context.Background()
+		infraEnv := &aiv1beta1.InfraEnv{}
+
+		if err := r.Get(ctx, types.NamespacedName{Name: a.GetName(), Namespace: a.GetNamespace()}, infraEnv); err != nil {
+			return []reconcile.Request{}
+		}
+
+		// Don't queue any reconcile if the InfraEnv
+		// doesn't have the ISODownloadURL set yet.
+		if infraEnv.Status.ISODownloadURL == "" {
+			return []reconcile.Request{}
+		}
+
+		bmh, err := r.findBMHByInfraEnv(ctx, infraEnv)
```

Comment:
Don't we want this to return a slice of all the BMHs that match the InfraEnv, then transform those into `[]reconcile.Request` ?

Reply:
And this, my kids, is the reason why unit-tests are worth almost nothing and everything should have proper functional/integration testing! 

![giphy](https://user-images.githubusercontent.com/13816/116802258-60155c00-ab09-11eb-8e78-fe131bb8a79d.gif)

Great catch!

Reply:
oh, hold on! it's actually there L858 but it's collapse because GH diffs are....

Reply:
This code is finding the first BMH that matches and immediately returning a request just for it. Instead it should continue iterating BMHs, find all of them that match the InfraEnv, and return requests for all of them.